src/com/nisovin/magicspells/spells/targeted/MaterializeSpell.java

```diff
@@ -10,41 +14,126 @@
-	
+	private String blocktype;
+
+	//Pattern Configuration
+	private boolean usePattern;
+	private List<String> patterns;
+	private MagicMaterial[][] rowPatterns;
+	private boolean restartpatterneachrow;
+	private boolean randomizepattern;
+
+	//Cuboid Parameters
+	private String area;
+	private int height;
+	private double fallheight;
+
+	//Cubiod Variables;
+	private int rowSize;
+	private int columnSize;
+
+	//Cubiod Checks;
+	private boolean hasMiddle;
+
+	//Delays
+	private double blockdelay;
+
 	public MaterializeSpell(MagicConfig config, String spellName) {
 		super(config, spellName);
-		
-		String s = getConfigString("block-type", "1");
-		material = MagicSpells.getItemNameResolver().resolveBlock(s);
+
+		blocktype = getConfigString("block-type", "1");
+		material = MagicSpells.getItemNameResolver().resolveBlock(blocktype);
 		if (material == null) MagicSpells.error("Invalid block-type on materialize spell '" + internalName + '\'');
 		resetDelay = getConfigInt("reset-delay", 0);
 		falling = getConfigBoolean("falling", false);
 		applyPhysics = getConfigBoolean("apply-physics", true);
 		checkPlugins = getConfigBoolean("check-plugins", true);
 		playBreakEffect = getConfigBoolean("play-break-effect", true);
 		strFailed = getConfigString("str-failed", "");
+
+		usePattern = getConfigBoolean("use-pattern",false);
+		patterns = getConfigStringList("patterns", null);
+		restartpatterneachrow = getConfigBoolean("restart-pattern-each-row", false);
+		randomizepattern = getConfigBoolean("randomize-pattern", false);
+
+		area = getConfigString("area", "1x1");
+		height = getConfigInt("height", 1);
+		fallheight = getConfigDouble("fall-height", 0.5);
+
+		blockdelay = getConfigDouble("delay-per-block", 0.0);
+	}
+
+	public void initialize() {
+		super.initialize();
+
+		//First, lets split the "area" that was given.
+		String[] areaparts = area.split("x",2);
+
+		//Lets define the size of the row and column to form an shape array;
+		rowSize = Integer.parseInt(areaparts[0]);
+		columnSize = Integer.parseInt(areaparts[1]);
+
+		/*For this to work smoothly, we need to see if the shape array has a middle;
+		It becomes very complicated when working with shape arrays without a block as a geometrical middle
+		So unfortunately. Shape arrays without a block as its geomtrical center cannot be accepted.
+		3x2, 9x8. Basically, if the product of the length and width is even. Don't use it. */
+		hasMiddle = ((rowSize*columnSize - 1) % 2 ) == 0;
+
+		if (!hasMiddle && patterns != null) {
+			MagicSpells.error("MaterializeSpell " + internalName + " is using a shape array without a geometrical center! A single block will spawn instead.");
+		}
+
+		//If height is 0, the code ceases to function. Lets not have that.
+		if (height == 0) height = 1;
+
+		//After the reset-delay passes, we need to remove all the blocks that were materialized.
+		//We store them within "materials" and "rowPatterns" aswell
+		boolean ready;
+
+		materials = new HashSet();
```

Comment:
Should avoid raw types by constructing it as `new HashSet<>()` instead of `new HashSet()`.
Additionally, you may want to consider using an EnumSet instead of a HashSet since the EnumSet is highly optimized.